util/jsonpull.c

```diff
@@ -0,0 +1,974 @@
+ */
+
+#include "jsonpull.h"
+
+#include <assert.h>
+
+#define GVM_JSON_CHAR_EOF -1       ///< End of file
+#define GVM_JSON_CHAR_ERROR -2     ///< Error reading file
+#define GVM_JSON_CHAR_UNDEFINED -3 ///< Undefined state
+
+/**
+ * @brief Escapes a string according to the JSON or JSONPath standard
+ */
+gchar *
+gvm_json_string_escape (const char *string, gboolean single_quote)
+{
+  gchar *point;
+  if (string == NULL)
+    return NULL;
+
+  GString *escaped = g_string_sized_new (strlen (string));
+  for (point = (char *) string; *point != 0; point++)
+    {
+      unsigned char character = *point;
+
+      if ((character > 31) && (character != '\\')
+          && (single_quote || character != '\"')
+          && (!single_quote || character != '\''))
+        {
+          g_string_append_c (escaped, character);
+        }
+      else
+        {
+          g_string_append_c (escaped, '\\');
+          switch (*point)
+            {
+            case '\\':
+            case '\'':
+            case '\"':
+              g_string_append_c (escaped, *point);
+              break;
+            case '\b':
+              g_string_append_c (escaped, 'b');
+              break;
+            case '\f':
+              g_string_append_c (escaped, 'f');
+              break;
+            case '\n':
+              g_string_append_c (escaped, 'n');
+              break;
+            case '\r':
+              g_string_append_c (escaped, 'r');
+              break;
+            case '\t':
+              g_string_append_c (escaped, 't');
+              break;
+            default:
+              g_string_append_printf (escaped, "u%04x", character);
+            }
+        }
+    }
+  return g_string_free (escaped, FALSE);
+}
+
+/**
+ * @brief Creates a new JSON path element.
+ *
+ * @param[in]  parent_type  Type of the parent (array, object, none/root)
+ * @param[in]  depth        The depth in the document tree
+ *
+ * @return The newly allocated path element
+ */
+gvm_json_path_elem_t *
+gvm_json_pull_path_elem_new (gvm_json_pull_container_type_t parent_type,
+                             int depth)
+{
+  gvm_json_path_elem_t *new_elem = g_malloc0 (sizeof (gvm_json_path_elem_t));
+  new_elem->parent_type = parent_type;
+  new_elem->depth = depth;
+  return new_elem;
+}
+
+/**
+ * @brief Frees a JSON path element.
+ *
+ * @param[in]  elem  The element to free
+ */
+void
+gvm_json_pull_path_elem_free (gvm_json_path_elem_t *elem)
+{
+  g_free (elem->key);
+  g_free (elem);
+}
+
+/**
+ * @brief Initializes a JSON pull event data structure.
+ *
+ * @param[in]  event  The event structure to initialize
+ */
+void
+gvm_json_pull_event_init (gvm_json_pull_event_t *event)
+{
+  memset (event, 0, sizeof (gvm_json_pull_event_t));
+}
+
+/**
+ * @brief Resets a JSON pull event data structure for reuse.
+ *
+ * @param[in]  event  The event structure to reset
+ */
+void
+gvm_json_pull_event_reset (gvm_json_pull_event_t *event)
+{
+  cJSON_free (event->value);
```

Comment:
I guess this should also free `event->error_message`, in case the caller resets the event after an error.